File: src/bridges/codex/stop_hook.ts
/**
 * Agent Comms — Codex Stop hook.
 *
 * Drains the agent-comms delivery queue when Codex finishes a turn.
 * If pending messages exist, returns decision=block with the messages
 * as the reason, causing Codex to continue processing them immediately.
 *
 * Run via: npx agent-comms bridge codex-stop
 *
 * Requires [features] codex_hooks = true in config.toml.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { DeliveryEventSchema } from "../../core/types.js";
import type { DeliveryEvent } from "../../core/types.js";

const BUS_ROOT = path.join(os.homedir(), ".agents", "bus");
const IDENTITY_FILE = path.join(BUS_ROOT, "identity", "codex.json");

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readIdentity(): string | undefined {
  try {
    const parsed: unknown = JSON.parse(fs.readFileSync(IDENTITY_FILE, "utf-8"));
    if (isRecord(parsed) && typeof parsed.id === "string") {
      return parsed.id;
    }
    return undefined;
  } catch {
    return undefined;
  }
}

function drainDelivery(agentId: string): DeliveryEvent[] {
  const deliveryDir = path.join(BUS_ROOT, "delivery", agentId);
  try {
    const files = fs
      .readdirSync(deliveryDir)
      .filter((f) => f.endsWith(".json"))
      .sort();
    const events: DeliveryEvent[] = [];
    for (const file of files) {
      const fp = path.join(deliveryDir, file);
      try {
        const parsed: unknown = JSON.parse(fs.readFileSync(fp, "utf-8"));
        events.push(DeliveryEventSchema.parse(parsed));
        fs.unlinkSync(fp);
      } catch {
        fs.unlinkSync(fp);
      }
    }
    events.sort((a, b) => {
      const ta =
        a.type === "room_message" || a.type === "dm" ? a.message.timestamp : "";
      const tb =
        b.type === "room_message" || b.type === "dm" ? b.message.timestamp : "";
      return ta.localeCompare(tb);
    });
    return events;
  } catch {
    return [];
  }
}

function formatEvent(event: DeliveryEvent): string {
  switch (event.type) {
    case "room_message":
      return `[${event.message.room}] ${event.message.from}: ${event.message.content}`;
    case "dm":
      return `DM from ${event.message.from}: ${event.message.content}`;
    case "room_invite":
      return `Invited to room ${event.room} by ${event.from}`;
    case "member_joined":
      return `${event.agent} joined ${event.room}`;
    case "member_left":
      return `${event.agent} left ${event.room}`;
  }
}

export function run(): void {
  const chunks: string[] = [];
  process.stdin.setEncoding("utf-8");
  process.stdin.on("data", (chunk: string) => chunks.push(chunk));
  process.stdin.on("end", () => {
    try {
      JSON.parse(chunks.join(""));
    } catch {
      /* no input, fine */
    }

    const agentId = readIdentity();
    if (!agentId) process.exit(0);

    const events = drainDelivery(agentId);
    if (events.length === 0) process.exit(0);

    const lines = events.map(formatEvent);
    const message =
      "📬 Agent Comms pending messages:\n" +
      lines.map((l) => `- ${l}`).join("\n");

    process.stdout.write(
      JSON.stringify({ decision: "block", reason: message }, null, 2) + "\n",
    );
  });
}

File: src/bridges/codex/stop_hook.py
#!/usr/bin/env python3
"""
Agent Bus — Codex Stop hook.

Drains the agent-bus delivery queue when Codex finishes a turn.
If pending messages exist, returns decision=block with the messages
as the reason, causing Codex to continue processing them immediately.

Install in ~/.codex/hooks.json:

{
  "hooks": {
    "Stop": [
      {
        "hooks": [
          {
            "type": "command",
            "command": "python3 ~/Developer/agent-bus/src/bridges/codex/stop_hook.py",
            "timeout": 5
          }
        ]
      }
    ]
  }
}

Requires [features] codex_hooks = true in config.toml.
"""

import json
import os
import sys
from pathlib import Path

BUS_ROOT = Path(os.path.expanduser("~/.agents/bus"))
IDENTITY_FILE = BUS_ROOT / "identity.json"


def read_identity() -> str | None:
    """Read the agent's bus identity."""
    try:
        data = json.loads(IDENTITY_FILE.read_text(encoding="utf-8"))
        return data.get("id")
    except (FileNotFoundError, json.JSONDecodeError, KeyError):
        return None


def drain_delivery(agent_id: str) -> list[dict]:
    """Drain and return all pending delivery events."""
    delivery_dir = BUS_ROOT / "delivery" / agent_id
    if not delivery_dir.is_dir():
        return []

    events: list[dict] = []
    for file in sorted(delivery_dir.glob("*.json")):
        try:
            event = json.loads(file.read_text(encoding="utf-8"))
            events.append(event)
            file.unlink()
        except (json.JSONDecodeError, OSError):
            # Remove malformed files
            file.unlink(missing_ok=True)

    # Sort by timestamp if available
    def get_ts(e: dict) -> str:
        msg = e.get("message", {})
        return msg.get("timestamp", "") if isinstance(msg, dict) else ""

    events.sort(key=get_ts)
    return events


def format_event(event: dict) -> str:
    """Format a delivery event as human-readable text."""
    event_type = event.get("type", "")
    if event_type == "room_message":
        msg = event.get("message", {})
        return f"[{msg.get('room', '?')}] {msg.get('from', '?')}: {msg.get('content', '')}"
    elif event_type == "dm":
        msg = event.get("message", {})
        return f"DM from {msg.get('from', '?')}: {msg.get('content', '')}"
    elif event_type == "room_invite":
        return f"Invited to room {event.get('room', '?')} by {event.get('from', '?')}"
    elif event_type == "member_joined":
        return f"{event.get('agent', '?')} joined {event.get('room', '?')}"
    elif event_type == "member_left":
        return f"{event.get('agent', '?')} left {event.get('room', '?')}"
    return f"Unknown event: {event_type}"


def main() -> None:
    # Read the Codex hook input from stdin
    hook_input = json.load(sys.stdin)

    agent_id = read_identity()
    if agent_id is None:
        # Not registered — nothing to do
        sys.exit(0)

    events = drain_delivery(agent_id)
    if not events:
        # No pending messages — let Codex stop normally
        sys.exit(0)

    # Format messages and block Codex's stop
    lines = [format_event(e) for e in events]
    message = "📬 Agent Bus pending messages:\n" + "\n".join(f"- {line}" for line in lines)

    result = {
        "decision": "block",
        "reason": message,
    }
    json.dump(result, sys.stdout, indent=2)
    print()  # trailing newline


if __name__ == "__main__":
    main()
